test/unit/sharding_key_test.lua

```diff
@@ -0,0 +1,104 @@
+local t = require('luatest')
```

Comment:
Regarding commit 'Support custom sharding key in replace, insert, upsert'.

I feel myself a bit worried without a test that actually calls insert/replace/upsert and checks where the tuple appears (on which replicaset) plus bucket_id within the tuple. Sketchy:

```lua
-- XXX: Ensure that the following statements are true (I didn't do
-- that).
--
-- The test leans on the following facts:
--
-- 1. bucket_id({123}) ~= bucket_id({'William'})
-- 2. A replicaset that owns bucket_id({123}) is different from one that owns
--    bucket_id({'William'}).
local object = {id = 123, name = 'William'}
--                  ^^^^^^^^  ^^^^^^^^^^^^^^^^
--                 primary key  sharding key
--              that's defined in the ddl module
local exp_bucket_id = vshard.router.bucket_id_strcrc32({'William'})
local exp_tuple = {123, 'William', exp_bucket_id}

-- Verify return value (mainly assigned bucket_id).
local res, err = g.cluster.main_server.net_box:call('crud.insert_object', {'some_space_name', src_object})
t.assert_nil(err)
t.assert_equals(res.rows[1], exp_tuple)

-- Verify stored tuple.
local stored_tuple = g.cluster.main_server.net_box:call('vshard.router.call', {exp_bucket_id,
    'box.space.some_space_name:get', {123}})
t.assert_equals(stored_tuple, exp_tuple)
```

Reply:
I see now, the 'Support custom sharding key in delete, update and get' commit adds tests, where we check for particular bucket_id in inserted tuples.

So we check that correct bucket_id is inserted into the tuple. Good. However we don't check that correct bucket_id is used for `vshard.router.call()` within crud. I'll demonstrate.

First, I fixed the problem with non-unique index, otherwise `test_non_unique_index` fails for me (see the link):

```diff
commit 95336d873f5cb69abc49d5fe57aee83872e023df
Author: Alexander Turenko <alexander.turenko@tarantool.org>
Date:   Sat Nov 13 10:28:56 2021 +0300

    FIXUP: fix select for non-pk sharding key for non-unique index
    
    See https://github.com/tarantool/crud/pull/181#discussion_r707761937

diff --git a/crud/select/plan.lua b/crud/select/plan.lua
index ca355ad..692905c 100644
--- a/crud/select/plan.lua
+++ b/crud/select/plan.lua
@@ -235,11 +235,6 @@ function select_plan.new(space, conditions, opts)
         sharding_key = extract_sharding_key_from_scan_value(scan_value, scan_index, sharding_index)
     end
 
-    if sharding_key ~= nil and opts.force_map_call ~= true then
-        total_tuples_count = 1
-        scan_iter = box.index.REQ
-    end
-
     local plan = {
         conditions = conditions,
         space_name = space_name,
diff --git a/test/unit/select_plan_test.lua b/test/unit/select_plan_test.lua
index 026c3e7..cc0dee3 100644
--- a/test/unit/select_plan_test.lua
+++ b/test/unit/select_plan_test.lua
@@ -159,7 +159,6 @@ g.test_is_scan_by_full_sharding_key_eq = function()
 
     t.assert_equals(err, nil)
 
-    t.assert_equals(plan.total_tuples_count, 1)
     t.assert_equals(plan.sharding_key, {15})
 
     -- id is a part of scan index
@@ -173,7 +172,6 @@ g.test_is_scan_by_full_sharding_key_eq = function()
 
     t.assert_equals(plan.index_id, 3) -- index name_id is used
     t.assert_equals(plan.scan_value, {'Ivan', 11})
-    t.assert_equals(plan.total_tuples_count, 1)
     t.assert_equals(plan.sharding_key, {11})
 
     -- other index is first
@@ -221,7 +219,6 @@ g.test_is_scan_by_full_sharding_key_eq = function()
 
     t.assert_equals(err, nil)
 
-    t.assert_equals(plan.total_tuples_count, 1)
     t.assert_equals(plan.sharding_key, {1, 0})
 end
```

Second, I found a problem with schema reloading. I worked it around, will file an issue soon.

```diff
commit dcf1aafca73dff3a1594f1e720853f596898f875
Author: Alexander Turenko <alexander.turenko@tarantool.org>
Date:   Sat Nov 13 10:30:44 2021 +0300

    update: fix stale schema problem
    
    I'll file an issue a bit later.

diff --git a/crud/update.lua b/crud/update.lua
index 587068c..2942e76 100644
--- a/crud/update.lua
+++ b/crud/update.lua
@@ -103,6 +103,18 @@ local function call_update_on_router(space_name, key, user_operations, opts)
     end
 
     local bucket_id = sharding.key_get_bucket_id(sharding_key, opts.bucket_id)
+
+    -- We should use a space object from the same connection as
+    -- we use for the request itself. Otherwise we can use a stale
+    -- format to generate rows metadata.
+    --
+    -- XXX: This fix is fast written and may have problems. IOW, I
+    -- didn't think much about it.
+    space, err = utils.get_space(space_name, {vshard.router.route(bucket_id)})
+    if err ~= nil then
+        return nil, UpdateError:new("Failed to get space %q: %s", space_name, err), true
+    end
+
     local call_opts = {
         mode = 'write',
         timeout = opts.timeout,
```

Third, let's break bucket_id calculation for `vshard.router.call()`.

```diff
commit 2720e7e70094107e3a1e3474498866a002007afd (HEAD -> tmp)
Author: Alexander Turenko <alexander.turenko@tarantool.org>
Date:   Sat Nov 13 10:31:37 2021 +0300

    DO NOT MERGE: break bucket_id calculation
    
    We initialize vshard with 3000 buckets in tests, so
    `bucket_id = (bucket_id + 1500) % 3000` will definitely call a wrong
    replicaset.
    
    But all tests succeeds. That's suspicious.

diff --git a/crud/common/call.lua b/crud/common/call.lua
index c5b94c9..08a2629 100644
--- a/crud/common/call.lua
+++ b/crud/common/call.lua
@@ -130,6 +130,7 @@ function call.single(bucket_id, func_name, func_args, opts)
 
     local timeout = opts.timeout or call.DEFAULT_VSHARD_CALL_TIMEOUT
 
+    bucket_id = (bucket_id + 1500) % 3000
     local res, err = vshard.router[vshard_call_name](bucket_id, func_name, func_args, {
         timeout = timeout,
     })
diff --git a/crud/select/compat/common.lua b/crud/select/compat/common.lua
index 2e23aa2..a39bd3b 100644
--- a/crud/select/compat/common.lua
+++ b/crud/select/compat/common.lua
@@ -7,6 +7,7 @@ local SELECT_FUNC_NAME = '_crud.select_on_storage'
 local DEFAULT_BATCH_SIZE = 100
 
 local function get_replicasets_by_sharding_key(bucket_id)
+    bucket_id = (bucket_id + 1500) % 3000
     local replicaset, err = vshard.router.route(bucket_id)
     if replicaset == nil then
         return nil, GetReplicasetsError:new("Failed to get replicaset for bucket_id %s: %s", bucket_id, err.err)
```

After that *all* tests succeeded, wow.

Well, I verified it on top of your branch and it does not contain recent storage sensitive tests from PR #222 — they would not pass (I hope).

IOW, we can have a problem in crud that will affect crud.insert() and crud.get() both in the way that will make illusion of correctness if we'll use only those two. But any other vshard user will see an incorrect picture. Or, say, a user of another version of crud (which is not affected by the problem).

Generally speaking, I prefer to don't use a product to prepare data for testing the product.

----

If you wonder, why I raised it within this PR and don't consider it as some irrelevant problem: in this PR we revisit bucket id calculation algorithm. Before it was quite obvious and based on pk. Now there are more options and so more questions to testing regarding ability to reveal problems around bucket id calculation.

----

However, well, if it'll not be so simple as I guess in my sketchy code (in the first message in this discussion), I don't mind to postpone such testing. It is up to you, Sergey.

Reply:
(Filed #236 regarding the problem that I worked around in the second patch above.)

Reply:
Okey, I got it. We need to make tests more reliable.

I have updated tests for select(): replaced common check to specific ones, test_select() used pk instead of secondary sharding key:
```diff
--- a/test/integration/ddl_sharding_key_test.lua
+++ b/test/integration/ddl_sharding_key_test.lua
@@ -64,14 +64,14 @@ pgroup.test_select = function(g)
     t.assert_not_equals(result, nil)
     t.assert_equals(#result.rows, 1)
 
-    local conditions = {{'==', 'id', 2}}
+    local conditions = {{'==', 'name', 'Ivan'}}
 
     local result, err = g.cluster.main_server.net_box:call('crud.select', {
         'customers_name_key', conditions,
     })
 
     t.assert_equals(err, nil)
-    t.assert_not_equals(result, nil)
+    t.assert_equals(result.rows, {{2, 1366, "Ivan", 20}})
     t.assert_equals(#result.rows, 1)
 end
 
@@ -139,7 +139,7 @@ pgroup.test_select_secondary_idx = function(g)
     })
 
     t.assert_equals(err, nil)
-    t.assert_not_equals(result, nil)
+    t.assert_equals(result.rows, {{2, 1366, "Ivan", 20}})
     t.assert_equals(#result.rows, 1)
 end
 ```

Now setup before testing made using non-crud operations (I use box operations on storages).

<details>
  <summary>Full patch</summary>

```diff
diff --git a/test/integration/ddl_sharding_key_test.lua b/test/integration/ddl_sharding_key_test.lua
index 10b24d6..aa0b7a6 100644
--- a/test/integration/ddl_sharding_key_test.lua
+++ b/test/integration/ddl_sharding_key_test.lua
@@ -62,12 +62,10 @@ local function check_get(g, space_name, id, name)
     t.assert_not_equals(result, nil)
 end
 
-pgroup.test_insert_object_get = function(g)
-    -- insert_object
+pgroup.test_insert_object = function(g)
     local result, err = g.cluster.main_server.net_box:call(
-        'crud.insert_object', {'customers_name_key', {id = 1, name = 'Fedor', age = 59}})
+        'crud.insert_object', {'customers_name_key', {id = 1, name = 'Augustus', age = 48}})
 
-    t.assert_equals(err, nil)
     t.assert_equals(result.metadata, {
         {is_nullable = false, name = 'id', type = 'unsigned'},
         {is_nullable = false, name = 'bucket_id', type = 'unsigned'},
@@ -75,29 +73,27 @@ pgroup.test_insert_object_get = function(g)
         {is_nullable = false, name = 'age', type = 'number'},
     })
     local objects = crud.unflatten_rows(result.rows, result.metadata)
-    t.assert_equals(objects, {{id = 1, bucket_id = 86, name = 'Fedor', age = 59}})
-
-    -- get
-    local result, err = g.cluster.main_server.net_box:call('crud.get', {'customers_name_key', {1, 'Fedor'}})
-
-    t.assert_equals(err, nil)
-    t.assert(result ~= nil)
-    local objects = crud.unflatten_rows(result.rows, result.metadata)
-    t.assert_equals(objects, {{age = 59, bucket_id = 86, id = 1, name = "Fedor"}})
-
-    -- insert_object again
-    local obj, err = g.cluster.main_server.net_box:call(
-        'crud.insert_object', {'customers_name_key', {id = 1, name = 'Alexander', age = 37}})
-
-    t.assert_equals(err, nil)
-    t.assert_equals(obj.rows, {{1, 1690, "Alexander", 37}})
+    t.assert_equals(objects, {{id = 1, bucket_id = 782, name = 'Augustus', age = 48}})
+
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({1, 'Augustus'})
+    ]], {'customers_name_key'})
+    -- There is a tuple on s2 that we inserted before using crud.insert_object().
+    t.assert_equals(result, {1, 782, 'Augustus', 48})
+
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({1, 'Augustus'})
+    ]], {'customers_name_key'})
+    -- There is no tuple on s1 that we inserted before using crud.insert_object().
+    t.assert_equals(result, nil)
 end
 
-pgroup.test_insert_get = function(g)
-    -- insert
+pgroup.test_insert = function(g)
+    -- Insert a tuple.
     local result, err = g.cluster.main_server.net_box:call(
         'crud.insert', {'customers_name_key', {2, box.NULL, 'Ivan', 20}})
-
     t.assert_equals(err, nil)
     t.assert_equals(result.metadata, {
         {is_nullable = false, name = 'id', type = 'unsigned'},
@@ -105,143 +101,199 @@ pgroup.test_insert_get = function(g)
         {is_nullable = false, name = 'name', type = 'string'},
         {is_nullable = false, name = 'age', type = 'number'},
     })
-    t.assert_equals(result.rows, {{2, 1366, "Ivan", 20}})
-
-    -- get
-    local result, err = g.cluster.main_server.net_box:call('crud.get', {'customers_name_key', {2, 'Ivan'}})
-
-    t.assert_equals(err, nil)
-    t.assert_not_equals(result, nil)
-    t.assert_equals(result.rows, {{2, 1366, "Ivan", 20}})
+    t.assert_equals(#result.rows, 1)
+    t.assert_equals(result.rows[1], {2, 1366, 'Ivan', 20})
+
+    -- There is a tuple on s2 that we inserted before using crud.insert().
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({2, 'Ivan'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {2, 1366, 'Ivan', 20})
+
+    -- There is no tuple on s1 that we inserted before using crud.insert().
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({2, 'Ivan'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, nil)
 end
 
 pgroup.test_update = function(g)
-    local tuple = {2, box.NULL, 'Ivan', 20}
+    -- bucket_id is 1366, storage is s-2
+    local tuple = {2, 1366, 'Ivan', 10}
+
+    -- Put tuple with to s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:insert(tuple)
+    ]], {'customers_name_key', tuple})
+    t.assert_not_equals(result, nil)
+
+    -- Put tuple with to s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:insert(tuple)
+    ]], {'customers_name_key', tuple})
+    t.assert_not_equals(result, nil)
 
+    -- Update a tuple.
     local update_operations = {
         {'+', 'age', 10},
     }
-
-    -- insert tuple
-    local result, err = g.cluster.main_server.net_box:call('crud.insert', {
-        'customers_name_key', tuple
-    })
-
-    t.assert_equals(err, nil)
-    t.assert_not_equals(result, nil)
-    t.assert_equals(#result.rows, 1)
-
     local result, err = g.cluster.main_server.net_box:call('crud.update', {
         'customers_name_key', {2, 'Ivan'}, update_operations,
     })
-
     t.assert_equals(err, nil)
-    t.assert_not_equals(result, nil)
     t.assert_equals(#result.rows, 1)
-
-    -- get
-    local result, err = g.cluster.main_server.net_box:call('crud.get', {'customers_name_key', {2, 'Ivan'}})
-
-    t.assert_equals(err, nil)
-    local objects = crud.unflatten_rows(result.rows, result.metadata)
-    t.assert_equals(objects, {{age = 30, bucket_id = 1366, id = 2, name = "Ivan"}})
+    t.assert_equals(result.rows, {{2, 1366, 'Ivan', 20}})
+
+    -- Tuple on s1 replicaset was not updated.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({2, 'Ivan'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {2, 1366, 'Ivan', 10})
+
+    -- Tuple on s2 replicaset was updated.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({2, 'Ivan'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {2, 1366, 'Ivan', 20})
 end
 
 pgroup.test_delete = function(g)
-    local tuple = {2, box.NULL, 'Ivan', 20}
-
-    -- insert tuple
-    local result, err = g.cluster.main_server.net_box:call('crud.insert', {
-        'customers_name_key', tuple
-    })
-
-    t.assert_equals(err, nil)
+    -- bucket_id is 596, storage is s-2
+    local tuple = {7, 596, 'Dimitrion', 20}
+
+    -- Put tuple to s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:insert(tuple)
+    ]], {'customers_name_key', tuple})
     t.assert_not_equals(result, nil)
-    t.assert_equals(#result.rows, 1)
-
-    local result, err = g.cluster.main_server.net_box:call('crud.delete', {
-        'customers_name_key', {2, 'Ivan'},
-    })
 
-    t.assert_equals(err, nil)
+    -- Put tuple to s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:insert(tuple)
+    ]], {'customers_name_key', tuple})
     t.assert_not_equals(result, nil)
 
-    local result, err = g.cluster.main_server.net_box:call('crud.get', {
-        'customers_name_key', {2, 'Ivan'}
+    -- Delete tuple.
+    local result, err = g.cluster.main_server.net_box:call('crud.delete', {
+        'customers_name_key', {7, 'Dimitrion'},
     })
-
     t.assert_equals(err, nil)
-    t.assert_not_equals(result, nil)
-    t.assert_equals(#result.rows, 0)
+    t.assert_equals(#result.rows, 1)
+    t.assert_equals(result.rows[1], {7, 596, 'Dimitrion', 20})
+
+    -- There is a tuple on s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({7, 'Dimitrion'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {7, 596, 'Dimitrion', 20})
+
+    -- There is no tuple on s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({7, 'Dimitrion'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, nil)
 end
 
-pgroup.test_insert = function(g)
-    local tuple = {2, box.NULL, 'Ivan', 20}
-
-    -- insert
-    local result, err = g.cluster.main_server.net_box:call('crud.insert', {
-        'customers_name_key', tuple,
-    })
+pgroup.test_replace = function(g)
+    -- bucket_id is 596, storage is s-2
+    local tuple = {7, 596, 'Dimitrion', 20}
+
+    -- Put tuple to s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:insert(tuple)
+    ]], {'customers_name_key', tuple})
+    t.assert_not_equals(result, nil)
 
-    t.assert_equals(err, nil)
+    -- Put tuple to s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:insert(tuple)
+    ]], {'customers_name_key', tuple})
     t.assert_not_equals(result, nil)
-    t.assert_equals(result.rows[1], {2, 1366, "Ivan", 20})
 
-    check_get(g, 'customers_name_key', 2, 'Ivan')
-end
+    local tuple = {7, box.NULL, 'Augustus', 21}
 
-pgroup.test_replace = function(g)
-    local tuple = {2, box.NULL, 'Jane', 21}
-
-    -- replace
+    -- Replace a tuple.
     local result, err = g.cluster.main_server.net_box:call('crud.replace', {
         'customers_name_key', tuple
     })
-
     t.assert_equals(err, nil)
     t.assert_not_equals(result, nil)
-    t.assert_equals(result.rows[1], {2, 1538, "Jane", 21})
+    t.assert_equals(#result.rows, 1)
+    t.assert_equals(result.rows[1], {7, 782, 'Augustus', 21})
 
-    check_get(g, 'customers_name_key', 2, 'Jane')
+    -- There is no replaced tuple on s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({7, 'Augustus'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, nil)
+
+    -- There is replaced tuple on s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({7, 'Augustus'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {7, 782, 'Augustus', 21})
 end
 
 pgroup.test_replace_object = function(g)
-    -- get
-    local result, err = g.cluster.main_server.net_box:call('crud.get', {'customers_name_key', {44, 'Ivan'}})
+    -- bucket_id is 596, storage is s-2
+    local tuple = {8, 596, 'Dimitrion', 20}
+
+    -- Put tuple to s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:insert(tuple)
+    ]], {'customers_name_key', tuple})
+    t.assert_not_equals(result, nil)
 
-    t.assert_equals(err, nil)
-    t.assert_equals(result.metadata, {
-        {is_nullable = false, name = 'id', type = 'unsigned'},
-        {is_nullable = false, name = 'bucket_id', type = 'unsigned'},
-        {is_nullable = false, name = 'name', type = 'string'},
-        {is_nullable = false, name = 'age', type = 'number'},
-    })
-    t.assert_equals(#result.rows, 0)
+    -- Put tuple to s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:insert(tuple)
+    ]], {'customers_name_key', tuple})
+    t.assert_not_equals(result, nil)
 
-    -- replace_object
+    -- Replace an object.
     local result, err = g.cluster.main_server.net_box:call(
-        'crud.replace_object', {'customers_name_key', {id = 44, name = 'John Doe', age = 25}})
-
+        'crud.replace_object', {'customers_name_key', {id = 8, name = 'John Doe', age = 25}})
     t.assert_equals(err, nil)
     local objects = crud.unflatten_rows(result.rows, result.metadata)
-    t.assert_equals(objects, {{id = 44, bucket_id = 1035, name = 'John Doe', age = 25}})
+    t.assert_equals(objects, {{id = 8, bucket_id = 1035, name = 'John Doe', age = 25}})
 
-    -- replace_object
-    local result, err = g.cluster.main_server.net_box:call(
-        'crud.replace_object', {'customers_name_key', {id = 44, name = 'Jane Doe', age = 18}})
+    -- There is no replaced tuple on s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({8, 'John Doe'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, nil)
 
-    t.assert_equals(err, nil)
-    local objects = crud.unflatten_rows(result.rows, result.metadata)
-    t.assert_equals(objects, {{id = 44, bucket_id = 2194, name = 'Jane Doe', age = 18}})
+    -- There is replaced tuple on s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({8, 'John Doe'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {8, 1035, 'John Doe', 25})
 end
 
 pgroup.test_upsert_object = function(g)
-    -- upsert_object first time
+    -- Upsert an object first time.
     local result, err = g.cluster.main_server.net_box:call(
         'crud.upsert_object', {'customers_name_key', {id = 66, name = 'Jack Sparrow', age = 25}, {
              {'+', 'age', 25},
     }})
-
     t.assert_equals(#result.rows, 0)
     t.assert_equals(result.metadata, {
         {is_nullable = false, name = 'id', type = 'unsigned'},
@@ -251,43 +303,89 @@ pgroup.test_upsert_object = function(g)
     })
     t.assert_equals(err, nil)
 
-    -- get
-    local result, err = g.cluster.main_server.net_box:call('crud.get', {'customers_name_key', {66, 'Jack Sparrow'}})
+    -- There is a tuple on s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({66, 'Jack Sparrow'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {66, 2719, 'Jack Sparrow', 25})
 
-    t.assert_equals(err, nil)
-    local objects = crud.unflatten_rows(result.rows, result.metadata)
-    t.assert_equals(objects, {{id = 66, bucket_id = 2719, name = 'Jack Sparrow', age = 25}})
+    -- There is no tuple on s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({66, 'Jack Sparrow'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, nil)
 
-    -- upsert_object the same query second time when tuple exists
+    -- Upsert the same query second time when tuple exists.
     local result, err = g.cluster.main_server.net_box:call(
        'crud.upsert_object', {'customers_name_key', {id = 66, name = 'Jack Sparrow', age = 25}, {
             {'+', 'age', 25},
     }})
-
     t.assert_equals(#result.rows, 0)
     t.assert_equals(err, nil)
 
-    -- get
-    local result, err = g.cluster.main_server.net_box:call('crud.get', {'customers_name_key', {66, 'Jack Sparrow'}})
+    -- There is an updated tuple on s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({66, 'Jack Sparrow'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {66, 2719, 'Jack Sparrow', 50})
 
-    t.assert_equals(err, nil)
-    local objects = crud.unflatten_rows(result.rows, result.metadata)
-    t.assert_equals(objects, {{age = 50, bucket_id = 2719, id = 66, name = "Jack Sparrow"}})
+    -- There is no tuple on s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({66, 'Jack Sparrow'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, nil)
 end
 
 pgroup.test_upsert = function(g)
     local tuple = {1, box.NULL, 'John', 25}
 
-    -- upsert
+    -- Upsert an object first time.
     local result, err = g.cluster.main_server.net_box:call('crud.upsert', {
         'customers_name_key', tuple, {}
     })
-
     t.assert_equals(err, nil)
     t.assert_not_equals(result, nil)
     t.assert_equals(#result.rows, 0)
 
-    check_get(g, 'customers_name_key', 1, 'John')
+    -- There is a tuple on s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({1, 'John'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {1, 2699, 'John', 25})
+
+    -- There is no tuple on s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({1, 'John'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, nil)
+
+    -- Upsert the same query second time when tuple exists.
+    local result, err = g.cluster.main_server.net_box:call(
+       'crud.upsert_object', {'customers_name_key', {id = 1, name = 'John', age = 25}, {
+            {'+', 'age', 25},
+    }})
+    t.assert_equals(#result.rows, 0)
+    t.assert_equals(err, nil)
+
+    -- There is an updated tuple on s1 replicaset.
+    local result = g.cluster:server('s1-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({1, 'John'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, {1, 2699, 'John', 50})
+
+    -- There is no tuple on s2 replicaset.
+    local result = g.cluster:server('s2-master').net_box:eval([[
+        local space_name, tuple = ...
+        return box.space[space_name]:get({1, 'John'})
+    ]], {'customers_name_key'})
+    t.assert_equals(result, nil)
 end
 
 pgroup.test_select = function(g)
```

</details>

Reply:
Nice!

Reply:
Also moved tests to appropriate commits (in previous version of path commit with 'upsert, insert, replace' was introduced without integration tests).

Reply:
Added more tests (for example for get()).

Reply:
I left testcase that check edge cases as is. These testcases are negative and check error messages:
- `pgroup.test_delete_incomplete_sharding_key`
- `pgroup.test_get_incomplete_sharding_key`
- `pgroup.test_update_incomplete_sharding_key`
- `pgroup.test_get_secondary_idx`
- `pgroup.test_update_secondary_idx`
- `pgroup.test_delete_secondary_idx`

I think it is acceptable to use CRUD inserts there for setup.

Reply:
I would also leave a note for 'broken' inserting, where we insert a tuple with some bucket_id to an incorrect storage (that does not oen this bucket) to check later that crud does not touch it (with update).